packages/salesforcedx-vscode-soql/src/editor/soqlEditorInstance.ts

```diff
@@ -181,11 +182,23 @@ export class SOQLEditorInstance {
         break;
       }
       case MessageType.RUN_SOQL_QUERY: {
-        this.handleRunQuery().catch(() => {
-          channelService.appendLine(
-            `An error occurred while running the SOQL query.`
-          );
-        });
+        vscode.window
+          .withProgress(
+            {
+              cancellable: false,
+              location: vscode.ProgressLocation.Notification,
+              title: 'Running query'
+            },
+            () => {
+              return this.handleRunQuery();
+            }
+          )
+          .then(undefined, () => {
+            channelService.appendLine(
+              `An error occurred while running the SOQL query.`
+            );
+            this.runQueryDone();
```

Comment:
I haven't seen this pattern before with `then` instead of `catch`. What is the `undefined` represent? The return prevents the `then ` from running in the happy path?

Does this block only run if `handleRunQuery` throws? or some other situation?

Reply:
Yeah this is because of VSCode's goofy Thenable wrapper around standard Promise chains. For Thenable the `then` function's first argument is the `resolve` callback (here it's `undefined` because there's no more work to do) and the second argument is the `reject` callback.